pkg/remotewrite/metrics.go

```diff
@@ -13,37 +13,66 @@ import (
 
 // metricsStorage is an in-memory gather point for metrics
 type metricsStorage struct {
-	m map[string]metrics.Sample
+	m map[string]*metrics.Metric
 }
 
 func newMetricsStorage() *metricsStorage {
 	return &metricsStorage{
-		m: make(map[string]metrics.Sample),
+		m: make(map[string]*metrics.Metric),
 	}
 }
 
 // update modifies metricsStorage and returns updated sample
 // so that the stored metric and the returned metric hold the same value
-func (ms *metricsStorage) update(sample metrics.Sample, add func(current, s metrics.Sample) metrics.Sample) metrics.Sample {
-	if current, ok := ms.m[sample.Metric.Name]; ok {
-		if add == nil {
-			current.Metric.Sink.Add(sample)
-		} else {
-			current = add(current, sample)
+func (ms *metricsStorage) update(sample metrics.Sample, add func(*metrics.Metric, metrics.Sample)) *metrics.Metric {
+	m, ok := ms.m[sample.Metric.Name]
+	if !ok {
+		var sink metrics.Sink
+		switch sample.Metric.Type {
+		case metrics.Counter:
+			sink = &metrics.CounterSink{}
+		case metrics.Gauge:
+			sink = &metrics.GaugeSink{}
+		case metrics.Trend:
+			sink = &metrics.TrendSink{}
+		case metrics.Rate:
+			sink = &metrics.RateSink{}
+		default:
+			panic("the Metric Type is not supported")
 		}
-		current.Time = sample.Time // to avoid duplicates in timestamps
-		// Sometimes remote write endpoint throws an error about duplicates even if the values
-		// sent were different. By current observations, this is a hard to repeat case and
-		// potentially a bug.
-		// Related: https://github.com/prometheus/prometheus/issues/9210
-
-		ms.m[current.Metric.Name] = current
-		return current
+
+		m = &metrics.Metric{
+			Name:     sample.Metric.Name,
+			Type:     sample.Metric.Type,
+			Contains: sample.Metric.Contains,
+			Sink:     sink,
+		}
+
+		ms.m[m.Name] = m
+	}
+
+	// TODO: this is just avoiding duplicates with the previous
+	// Implement a better and complete solution
+	// maybe discard any timestamp < latest?
+	//
+	//current.Time = sample.Time // to avoid duplicates in timestamps
```

Comment:
I find just now the time for checking this comment. The Prometheus' timestamp generator function https://github.com/grafana/xk6-output-prometheus-remote/blob/78df11454191fd1da5742b05a0a3a99de773098a/pkg/remotewrite/prometheus.go#L101 that we are using for generating the timestamp has a millisecond precision that it could be very low if we consider our high concurrent world. If the remote write supports higher precision we could just use the `sample.Time.UnixNano()`, if not then we need to discard or wait for the new data modelling (hopefully available soon).

Reply:
Yes, I agree that improving precision could help with this particular problem but even if it's possible, I'm not sure about drawbacks it'll bring. E.g. do people actually need precision higher than ms? what else would it impact during processing? etc. Also, I believe precision question is valid for both Prometheus and k6 implementations, IIRC.
So IMO, precision is a multi-level problem here.

This change feels like a bit of an experiment to me :slightly_smiling_face: Definitely :+1: for making more experiments and figuring out when exactly duplicate timestamps happen.